pkg/stub/handler.go

```diff
@@ -0,0 +1,162 @@
+package stub
+
+import (
+	"context"
+
+	"github.com/integr8ly/tutorial-web-app-operator/pkg/apis/integreatly/v1alpha1"
+
+	"github.com/operator-framework/operator-sdk/pkg/sdk"
+	"github.com/sirupsen/logrus"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/integr8ly/tutorial-web-app-operator/pkg/apis/integreatly/openshift"
+	"github.com/openshift/api/template/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"fmt"
+	"github.com/operator-framework/operator-sdk/pkg/util/k8sutil"
+	"src/github.com/pkg/errors"
+	errors2 "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/dynamic"
+	corev1 "k8s.io/api/core/v1"
+)
+
+func NewHandler(m *Metrics, osClient openshift.OSClient, dynamicResourceClientFactory func(apiVersion, kind, namespace string) (dynamic.ResourceInterface, string, error)) sdk.Handler {
+	return &Handler{
+		metrics: m,
+		osClient: osClient,
+		dynamicResourceClientFactory: dynamicResourceClientFactory,
+	}
+}
+
+type Metrics struct {
+	operatorErrors prometheus.Counter
+}
+
+type Handler struct {
+	// Metrics example
+	metrics *Metrics
+	osClient openshift.OSClient
+	dynamicResourceClientFactory func(apiVersion, kind, namespace string) (dynamic.ResourceInterface, string, error)
+}
+
+func (h *Handler) Handle(ctx context.Context, event sdk.Event) error {
+	switch o := event.Object.(type) {
+	case *v1alpha1.WebApp:
+		if o.Status.Phase == v1alpha1.NoPhase {
+			logrus.Infof("Starting provisioning phase")
+			exts, err := h.processTemplate(o)
+			if err != nil {
+				logrus.Errorf("%v", err)
```

Comment:
error handling in operators is kind of interesting. I think if we can't handle the error we should return the error at this point. We could possibly add a status message with the error message if it is not already present. Once an error is seen we will get to retry the operation. If it fails a lot it get dropped from the queue.

Reply:
Do you think we should throw a fatal error in this case? Could introduce a new status phase such as "failed" and store the error in it cr

Reply:
Also a non related question, do you think we should use some kind of finalizer if to delete the web app resources if the cr gets deleted as well?

Reply:
Yes I think we should always have a finalizer. This way we can ensure we have seen the delete before the object is removed